### ISSUE TITLE
Add a toxic to simulate "connection reset"

### ISSUE THREAD

**Issue by holyjak** [Shopify/toxiproxy#238]
A problem I experience regularly (in Java) is the `java.net.SocketException: Connection reset [by peer]`. It would be awesome to have a toxic that can cause it. Thanks!

Some background: https://docs.oracle.com/javase/8/docs/technotes/guides/net/articles/connection_release.html

**Comment by boly38** [comment]
I'm also interested by this usecase.
It seems feasible in java [following this example](http://appcrawler.com/wordpress/2016/12/19/simulating-a-connection-reset-in-java/):
- step A) accepting a socket, and map a stream reader
- step B) brute force exit 

_client class connects, then sleeps two seconds before attempting to read the stream from the now defunct server socket_

Then, I dont know if it's easy to translate this sample/behaviour onto custom toxiproxy toxic.

**Comment by holyjak** [comment]
I do not understand it deeply enough to know how to simulate it but there must be a way to do something at the other end (toxiproxy) here to simulate "connection reset by peer" .

@boly38 What you describe is I think the other case, when the connection is not reset by the peer but within the JVM itself due to trying to read from a closed socket or something like that.

**Comment by jpittis** [comment]
Alrighty. I was wrong. We can simulate a connection reset by peer from an application.

Here's a Go script I wrote that triggers a connection reset by peer using `SetLinger(0)`.

[Link to executable Gist.](https://gist.github.com/jpittis/4357d817dc425ae99fbf719828ab1800)

Running the experiment looks like this:

```
$ go run experiment.go
2019/01/02 08:29:36 connect err: EOF
exit status 1
$ go run experiment.go rst
2019/01/02 08:29:40 connect err: read tcp 127.0.0.1:50361->127.0.0.1:4444: read: connection reset by peer
exit status 1
```

This sounds like something we could for sure support in Toxiproxy.

**Comment by jpittis** [comment]
I don't have time to work on this. I may have time to suggest a way to implement it.

Until then, my only thoughts are that we'd need to figure out how to trigger a `SetLinger(0)` directly on the connecting client before close.

- [I'm pretty sure the client is closed here.](https://github.com/Shopify/toxiproxy/blob/master/link.go#L84)
- What's hard about this problem is toxics don't currently have direct access to the socket layer of abstraction so we'd need to find a way to properly abstract `SoLinger` before close.

**Comment by chaosbox** [comment]
@jpittis 
Created a PR to add this toxic, when you get a chance let us know what you think.

https://github.com/Shopify/toxiproxy/pull/247

**Comment by miry** [comment]
Released #247 in https://github.com/Shopify/toxiproxy/releases/tag/v2.2.0
